cpp/src/arrow/compute/exec/hash_join.cc

```diff
@@ -282,6 +279,125 @@ class HashJoinBasicImpl : public HashJoinImpl {
+          int key_idx = to_key.get(i);
+          concatenated.values.push_back(key.values[key_idx]);
+        } else if (to_pay.get(i) != SchemaProjectionMap::kMissingField) {
+          int pay_idx = to_pay.get(i);
+          concatenated.values.push_back(payload.values[pay_idx]);
+        }
+      }
+    };
+
+    SchemaProjectionMap left_to_key = schema_mgr_->proj_maps[0].map(
+        HashJoinProjection::FILTER, HashJoinProjection::KEY);
+    SchemaProjectionMap left_to_pay = schema_mgr_->proj_maps[0].map(
+        HashJoinProjection::FILTER, HashJoinProjection::PAYLOAD);
+    SchemaProjectionMap right_to_key = schema_mgr_->proj_maps[1].map(
+        HashJoinProjection::FILTER, HashJoinProjection::KEY);
+    SchemaProjectionMap right_to_pay = schema_mgr_->proj_maps[1].map(
+        HashJoinProjection::FILTER, HashJoinProjection::PAYLOAD);
+
+    AppendFields(left_to_key, left_to_pay, left_key, left_payload);
+    AppendFields(right_to_key, right_to_pay, right_key, right_payload);
+
+    ARROW_ASSIGN_OR_RAISE(Datum mask,
+                          ExecuteScalarExpression(filter_, concatenated, ctx_));
+
+    size_t num_probed_rows = match.size() + no_match.size();
+    if (mask.is_scalar()) {
+      const auto& mask_scalar = mask.scalar_as<BooleanScalar>();
+      if (mask_scalar.is_valid && mask_scalar.value) {
+        // All rows passed, nothing left to do
+        return Status::OK();
+      } else {
+        // Nothing passed, no_match becomes everything
+        no_match.resize(num_probed_rows);
+        std::iota(no_match.begin(), no_match.end(), 0);
+        match_left.clear();
+        match_right.clear();
+        match.clear();
+        return Status::OK();
+      }
+    }
+    ARROW_DCHECK(mask.array()->offset == 0);
+    ARROW_DCHECK(mask.array()->length == static_cast<int64_t>(match_left.size()));
+    const uint8_t* nulls = mask.array()->buffers[0]->data();
+    const uint8_t* comparisons = mask.array()->buffers[1]->data();
+    size_t num_rows = match_left.size();
+
+    match.clear();
+    no_match.clear();
+
+    int32_t match_idx = 0;  // current size of new match_left
+    int32_t irow = 0;       // index into match_left
+    for (int32_t curr_left = 0; static_cast<size_t>(curr_left) < num_probed_rows;
+         curr_left++) {
+      int32_t advance_to = static_cast<size_t>(irow) < num_rows
+                               ? match_left[irow]
+                               : static_cast<int32_t>(num_probed_rows);
+      while (curr_left < advance_to) {
+        no_match.push_back(curr_left++);
+      }
+      bool passed = false;
+      for (; static_cast<size_t>(irow) < num_rows && match_left[irow] == curr_left;
+           irow++) {
+        bool is_null = !BitUtil::GetBit(nulls, irow);
+        bool is_cmp_true = BitUtil::GetBit(comparisons, irow);
+        // We treat a null comparison result as false, like in SQL
+        if (!is_null && is_cmp_true) {
```

Comment:
Super minor nit:  `is_valid` might be clearer to read since you wouldn't have a double negation.